plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/KtlintPlugin.kt

```diff
@@ -156,18 +155,9 @@ open class KtlintPlugin : Plugin<Project> {
              */
             fun getPluginConfigureAction(): (Plugin<Any>) -> Unit = {
                 target.extensions.configure(BaseExtension::class.java) { ext ->
-                    ext.sourceSets { sourceSet ->
-                        sourceSet.all { androidSourceSet ->
-                            // Passing Callable, so returned FileCollection, will lazy evaluate it
-                            // only when task will need it.
-                            // Solves the problem of having additional source dirs in
-                            // current AndroidSourceSet, that are not available on eager
-                            // evaluation.
-                            createTasks(
-                                androidSourceSet.name,
-                                target.files(Callable { androidSourceSet.java.srcDirs })
-                            )
-                        }
+                    ext.variants.all { variant ->
+                        val files = variant.sourceSets.flatMap { sourceSet -> sourceSet.javaDirectories }
```

Comment:
```suggestion
                        val filesCallable = Callable { variant.sourceSets.flatMap { sourceSet -> sourceSet.javaDirectories } }
```

Reply:
This is actually not necessary thanks to `variants.all` above. There are sourceSet, flavor and buildType definitions in Android extension. At the point where a buildVariant is computed, all of them are already resolved.

Reply:
So a user can't add their own custom variant type?

Reply:
They can. `.all` is lazy and live loop over all variants. Default + custom variants arrive here together will additional custom sourceSets and sourceDirs inside them.

Reply:
Sure, you're looping over variants correctly, the problem you have is that what if you end up with user code in a scenario like this:

```kotlin
fun getPluginConfigureAction(): (Plugin<Any>) -> Unit = {
    target.extensions.configure(BaseExtension::class.java) { ext ->
        ext.variants.all { variant ->
            val files = variant.sourceSets.flatMap { sourceSet -> sourceSet.javaDirectories }
            createTasks(variant.name, target.files(files))

            // Somewhere off in user build code land later on...
            variant.addJavaSourceFoldersToModel(File(......))
        }
    }
}
```

Now, because of the eager evaluation of `variant.sourceSets.flatMap { sourceSet -> sourceSet.javaDirectories }` the file that is added later is completely missed.

Reply:
Hmmm, this is interesting. I will try this out to see what happens. The recommended way of adding source dirs is to use sourceSets extension on Android extension but we should change this still to be safe than sorry. 👍